src/algebraic_geometry/EllipticCurve.lean

```diff
@@ -33,80 +33,179 @@ Some more details of the construction can be found on pages 66-69 of
+  literature; we choose the sign used by the LMFDB. For more discussion, see
+  [the LMFDB page on discriminants](https://www.lmfdb.org/knowledge/show/ec.discriminant). -/
+@[simp] def EllipticCurve.Δ_aux {R : Type u} [comm_ring R] (a₁ a₂ a₃ a₄ a₆ : R) : R :=
+let b₂ : R := a₁ ^ 2 + 4 * a₂,
+    b₄ : R := 2 * a₄ + a₁ * a₃,
+    b₆ : R := a₃ ^ 2 + 4 * a₆,
+    b₈ : R := a₁ ^ 2 * a₆ + 4 * a₂ * a₆ - a₁ * a₃ * a₄ + a₂ * a₃ ^ 2 - a₄ ^ 2
+in  -b₂ ^ 2 * b₈ - 8 * b₄ ^ 3 - 27 * b₆ ^ 2 + 9 * b₂ * b₄ * b₆
+
 /-- The category of elliptic curves over `R` (note that this definition is only mathematically
-correct for certain rings, for example if `R` is a field or a PID). -/
-structure EllipticCurve (R : Type*) [comm_ring R] :=
-(a1 a2 a3 a4 a6 : R)
-(disc_unit : Rˣ)
-(disc_unit_eq : (disc_unit : R) = EllipticCurve.disc_aux a1 a2 a3 a4 a6)
+  correct for certain rings `R` with `Pic(R)[12] = 0`, for example if `R` is a field or a PID). -/
+structure EllipticCurve (R : Type u) [comm_ring R] :=
+(a₁ a₂ a₃ a₄ a₆ : R) (Δ : Rˣ) (Δ_eq : ↑Δ = EllipticCurve.Δ_aux a₁ a₂ a₃ a₄ a₆)
 
 namespace EllipticCurve
 
-instance : inhabited (EllipticCurve ℚ) := ⟨⟨0,0,1,-1,0, units.mk0 37 (by norm_num),
-  by simp only [units.coe_mk, disc_aux_def]; norm_num⟩⟩
+instance : inhabited (EllipticCurve ℚ) :=
+⟨⟨0, 0, 1, -1, 0, ⟨37, 37⁻¹, by norm_num1, by norm_num1⟩, show (37 : ℚ) = _ + _, by norm_num1 ⟩⟩
+
+variables {R : Type u} [comm_ring R] (E : EllipticCurve R)
+
+/-! ### Standard quantities -/
+
+section quantity
+
+/-- The `b₂` coefficient of an elliptic curve. -/
+@[simp] def b₂ : R := E.a₁ ^ 2 + 4 * E.a₂
+
+/-- The `b₄` coefficient of an elliptic curve. -/
+@[simp] def b₄ : R := 2 * E.a₄ + E.a₁ * E.a₃
+
+/-- The `b₆` coefficient of an elliptic curve. -/
+@[simp] def b₆ : R := E.a₃ ^ 2 + 4 * E.a₆
+
+/-- The `b₈` coefficient of an elliptic curve. -/
+@[simp] def b₈ : R :=
+E.a₁ ^ 2 * E.a₆ + 4 * E.a₂ * E.a₆ - E.a₁ * E.a₃ * E.a₄ + E.a₂ * E.a₃ ^ 2 - E.a₄ ^ 2
+
+lemma b_relation : 4 * E.b₈ = E.b₂ * E.b₆ - E.b₄ ^ 2 := by { simp, ring1 }
+
+/-- The `c₄` coefficient of an elliptic curve. -/
+@[simp] def c₄ : R := E.b₂ ^ 2 - 24 * E.b₄
+
+/-- The `c₆` coefficient of an elliptic curve. -/
+@[simp] def c₆ : R := -E.b₂ ^ 3 + 36 * E.b₂ * E.b₄ - 216 * E.b₆
+
+@[simp] lemma coe_Δ :
+  ↑E.Δ = -E.b₂ ^ 2 * E.b₈ - 8 * E.b₄ ^ 3 - 27 * E.b₆ ^ 2 + 9 * E.b₂ * E.b₄ * E.b₆ :=
+E.Δ_eq
+
+lemma c_relation : 1728 * ↑E.Δ = E.c₄ ^ 3 - E.c₆ ^ 2 := by { simp, ring1 }
+
+/-- The j-invariant of an elliptic curve, which is invariant under isomorphisms over `R`. -/
+@[simp] def j : R := ↑E.Δ⁻¹ * E.c₄ ^ 3
+
+end quantity
+
+/-! ### `2`-torsion polynomials -/
+
+section torsion_polynomial
+
+variables (A : Type v) [comm_ring A] [algebra R A]
+
+/-- The polynomial whose roots over a splitting field of `R` are the `2`-torsion points of the
+  elliptic curve when `R` is a field of characteristic different from `2`, and whose discriminant
+  happens to be a multiple of the discriminant of the elliptic curve. -/
```

Comment:
How far are we from showing that this does what it says on the tin?

Reply:
I already have this in the `mordell_weil` branch, but it's assuming the group law, whose associativity we can't prove. However, just defining addition law is sorry-free, so the torsion points can be defined as a set, and this is basically done.